proposals/scalable-robust-operator.md

```diff
@@ -0,0 +1,284 @@
+The latest release includes a v1alpha2 API and controller.
+A v1 is under development. Something to highlight in this new version is the
+replacement of the Job and StatefulSet by plain Pods, with the intent of
+[tracking running Pods](https://github.com/kubeflow/mpi-operator/issues/201#issuecomment-827837831).
+
+An MPIJob CRD describes the Job. Important fields include:
+- The workers template
+- The number of workers
+- The launcher template, which should have a `mpirun` command.
+
+The images are expected to have the MPI implementation binaries (such as
+OpenMPI, MPICH or Intel MPI) the user’s MPI executable.
+
+A controller processes the MPIJob, starting a Job with the following steps:
+1. Creates ConfigMap, which contains:
+  - A script `kubexec.sh` that wraps `kubectl exec` and is used in replacement
+    of `ssh`. This script, before executing the command provided by `mpirun`,
+    transfers a file containing a mapping of pod names to IPs and appends it to
+    the worker’s `/etc/hosts`.
+    
+    Note: The v1 controller no longer copies the pod-to-IP mapping file. The
+    OpenMPI implementation does the [routing](https://www.open-mpi.org/faq/?category=tcp#tcp-routability-1.3).
+  - The `hostfile` for `mpirun`, listing the worker pod names and number of
+    slots (which could be the number of CPUs or GPUs). This list is built
+    programmatically.
+2. Creates a ServiceAccount+Role+RoleBinding for the launcher, which allow it
+  to:
+  - get/list/watch on Pods
+  - do pods/exec
+  This allows the launcher Pod to obtain details of the worker Pods and start
+  the process managers on them.
+3. If configured, it creates a Volcano PodGroup
+4. Creates a StatefulSet for workers (plain pods in v1). The Pod template
+  includes:
+  - mounts for the ConfigMap.
+  - `sleep` as command.
+5. Creates launcher Job (plain pod in v1). The Pod template includes:
+  - An init container, `kubectl-delivery`, described below.
+  - Environment variables for:
+    - replacing `ssh` for `kubexec.sh`
+    - the `hostfile` location
+  - Volumes for:
+    - The ConfigMap
+    - Sharing files from `kubectl-delivery`
+
+The launcher Job, as previously mentioned, contains an init container:
+`kubectl-delivery`. This is a Kubernetes controller that watches pods. It does
+the following:
+
+1. Copy kubectl from the image into the volume shared with the main container.
+2. Wait for all Pods in the hostfile to be running
+3. Generates a file mapping pod name to IP, in `/etc/hosts` format.
+
+To update the status of an MPIJob, the controller uses the status of the
+launcher Job. That is, when the launcher Job fails or succeeds, it’s status is
+copied to the MPIJob. In v1, it bases the status on the termination condition of
+the Pod.
+
+### Analysis
+
+The above architecture for MPI Jobs has the following problems:
+
+- It doesn’t scale as we increase the number of workers, belonging to the same
```

Comment:
I believe this issue is addressed in v1 controller when we enable horovod elastic work on mpi-operator. https://github.com/kubeflow/mpi-operator/blob/master/proposals/elastic-horovod.md

Besides, I'm not sure what it means 'workers belonging to the same or different jobs'. What is the user case to increase the number of workers for a *different* job?

Reply:
The sentence perhaps needs some rephrasing, I think the point here is that the current setup doesn't scale when increasing the number of workers of a job (e.g., running a 1000 worker pod Job), or increasing the number of jobs (running 100s of MPI jobs). 

The current design puts a lot pressure on central component, the api server:

- `kubectl exec` runs through the API server, and so all jobs running on the same cluster will have to launch their workers using connections established through the api server, which is not scalable. This doesn't scale as we increase the total number of workers across all jobs running on the same cluster.
- pod cache syncs are expensive on large clusters, doing that for each job results in streaming a large amount of state from the api server to the launcher pod. This doesn't scale as we increase the number of jobs.